File: algorithms/genetic.js
/**
 * Genetic Algorithm for TSP
 * Compatible with both browser and Node.js
 * @see https://github.com/konard/sonar
 */

(function (root, factory) {
  if (typeof module === 'object' && module.exports) {
    const TSPUtils = require('./utils.js');
    module.exports = factory(TSPUtils);
  } else {
    root.GeneticAlgorithm = factory(root.TSPUtils);
  }
}(typeof self !== 'undefined' ? self : this, function (TSPUtils) {
  'use strict';

  const calculateTourLength = TSPUtils ? TSPUtils.calculateTourLength : function(tour, graph) {
    let totalDistance = 0;
    const n = tour.length;
    for (let i = 0; i < n; i++) {
      totalDistance += graph[tour[i]][tour[(i + 1) % n]];
    }
    return totalDistance;
  };

  /**
   * Create a random tour permutation
   * @param {number} n - Number of cities
   * @returns {Array<number>} Random tour
   */
  function randomTour(n) {
    const tour = Array.from({ length: n }, (_, i) => i);
    // Fisher-Yates shuffle
    for (let i = n - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [tour[i], tour[j]] = [tour[j], tour[i]];
    }
    return tour;
  }

  /**
   * Order Crossover (OX) operator
   * @param {Array<number>} parent1 - First parent tour
   * @param {Array<number>} parent2 - Second parent tour
   * @returns {Array<number>} Child tour
   */
  function crossover(parent1, parent2) {
    const n = parent1.length;
    const start = Math.floor(Math.random() * n);
    const end = start + Math.floor(Math.random() * (n - start));

    const child = new Array(n).fill(-1);
    const used = new Set();

    // Copy segment from parent1
    for (let i = start; i <= end; i++) {
      child[i] = parent1[i];
      used.add(parent1[i]);
    }

    // Fill rest from parent2 in order
    let idx = (end + 1) % n;
    for (let i = 0; i < n; i++) {
      const city = parent2[(end + 1 + i) % n];
      if (!used.has(city)) {
        child[idx] = city;
        idx = (idx + 1) % n;
      }
    }

    return child;
  }

  /**
   * Mutation by swapping two random cities
   * @param {Array<number>} tour - Tour to mutate
   * @returns {Array<number>} Mutated tour
   */
  function mutate(tour) {
    const newTour = [...tour];
    const n = tour.length;
    const i = Math.floor(Math.random() * n);
    const j = Math.floor(Math.random() * n);
    [newTour[i], newTour[j]] = [newTour[j], newTour[i]];
    return newTour;
  }

  /**
   * Optimize TSP using a Genetic Algorithm
   * Evolutionary approach using selection, crossover, and mutation
   * Time complexity: O(populationSize * generations * n)
   *
   * @param {Array<Array<number>>} graph - Distance matrix
   * @param {number} n - Number of cities
   * @param {number} populationSize - Size of the population
   * @param {number} generations - Number of generations to evolve
   * @param {number} mutationRate - Probability of mutation (0-1)
   * @returns {Array<number>} Best tour found
   */
  function optimize(graph, n, populationSize = 50, generations = 100, mutationRate = 0.1) {
    // Initialize population
    let population = [];
    for (let i = 0; i < populationSize; i++) {
      population.push(randomTour(n));
    }

    for (let gen = 0; gen < generations; gen++) {
      // Calculate fitness (inverse of tour length)
      const fitnesses = population.map(tour => 1 / calculateTourLength(tour, graph));
      const totalFitness = fitnesses.reduce((a, b) => a + b, 0);

      // Roulette wheel selection
      function select() {
        let r = Math.random() * totalFitness;
        for (let i = 0; i < populationSize; i++) {
          r -= fitnesses[i];
          if (r <= 0) return population[i];
        }
        return population[populationSize - 1];
      }

      // Create new population
      const newPopulation = [];

      // Elitism: keep the best individual
      let bestIdx = 0;
      for (let i = 1; i < populationSize; i++) {
        if (fitnesses[i] > fitnesses[bestIdx]) bestIdx = i;
      }
      newPopulation.push(population[bestIdx]);

      // Generate rest of population through selection and crossover
      while (newPopulation.length < populationSize) {
        const parent1 = select();
        const parent2 = select();
        let child = crossover(parent1, parent2);

        // Apply mutation with given probability
        if (Math.random() < mutationRate) {
          child = mutate(child);
        }

        newPopulation.push(child);
      }

      population = newPopulation;
    }

    // Return the best tour from final population
    let bestTour = population[0];
    let bestLength = calculateTourLength(bestTour, graph);

    for (const tour of population) {
      const length = calculateTourLength(tour, graph);
      if (length < bestLength) {
        bestTour = tour;
        bestLength = length;
      }
    }

    return bestTour;
  }

  return {
    optimize,
    randomTour,
    crossover,
    mutate
  };
}));

File: rust/src/genetic.rs
//! Genetic Algorithm for TSP

use crate::utils::calculate_tour_length;
use rand::Rng;
use std::collections::HashSet;

/// Create a random tour permutation using Fisher-Yates shuffle
fn random_tour(n: usize, rng: &mut impl Rng) -> Vec<usize> {
    let mut tour: Vec<usize> = (0..n).collect();
    for i in (1..n).rev() {
        let j = rng.gen_range(0..=i);
        tour.swap(i, j);
    }
    tour
}

/// Order Crossover (OX) operator
fn crossover(parent1: &[usize], parent2: &[usize], rng: &mut impl Rng) -> Vec<usize> {
    let n = parent1.len();
    let start = rng.gen_range(0..n);
    let end = start + rng.gen_range(0..n - start);

    let mut child = vec![usize::MAX; n];
    let mut used: HashSet<usize> = HashSet::new();

    // Copy segment from parent1
    for i in start..=end {
        child[i] = parent1[i];
        used.insert(parent1[i]);
    }

    // Fill rest from parent2 in order
    let mut idx = (end + 1) % n;
    for i in 0..n {
        let city = parent2[(end + 1 + i) % n];
        if !used.contains(&city) {
            child[idx] = city;
            idx = (idx + 1) % n;
        }
    }

    child
}

/// Mutation by swapping two random cities
fn mutate(tour: &[usize], rng: &mut impl Rng) -> Vec<usize> {
    let mut new_tour = tour.to_vec();
    let n = tour.len();
    let i = rng.gen_range(0..n);
    let j = rng.gen_range(0..n);
    new_tour.swap(i, j);
    new_tour
}

/// Optimize TSP using a Genetic Algorithm
/// Evolutionary approach using selection, crossover, and mutation
/// Time complexity: O(populationSize * generations * n)
pub fn optimize(
    graph: &[Vec<f64>],
    n: usize,
    population_size: usize,
    generations: usize,
    mutation_rate: f64,
) -> Vec<usize> {
    let mut rng = rand::thread_rng();

    // Initialize population
    let mut population: Vec<Vec<usize>> = (0..population_size)
        .map(|_| random_tour(n, &mut rng))
        .collect();

    for _ in 0..generations {
        // Calculate fitness (inverse of tour length)
        let fitnesses: Vec<f64> = population
            .iter()
            .map(|tour| 1.0 / calculate_tour_length(tour, graph))
            .collect();
        let total_fitness: f64 = fitnesses.iter().sum();

        // Roulette wheel selection
        let select = |rng: &mut rand::rngs::ThreadRng| -> &Vec<usize> {
            let mut r = rng.gen::<f64>() * total_fitness;
            for (i, &fitness) in fitnesses.iter().enumerate() {
                r -= fitness;
                if r <= 0.0 {
                    return &population[i];
                }
            }
            &population[population_size - 1]
        };

        // Create new population
        let mut new_population: Vec<Vec<usize>> = Vec::with_capacity(population_size);

        // Elitism: keep the best individual
        let best_idx = fitnesses
            .iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .map(|(i, _)| i)
            .unwrap_or(0);
        new_population.push(population[best_idx].clone());

        // Generate rest of population through selection and crossover
        while new_population.len() < population_size {
            let parent1 = select(&mut rng);
            let parent2 = select(&mut rng);
            let mut child = crossover(parent1, parent2, &mut rng);

            // Apply mutation with given probability
            if rng.gen::<f64>() < mutation_rate {
                child = mutate(&child, &mut rng);
            }

            new_population.push(child);
        }

        population = new_population;
    }

    // Return the best tour from final population
    let mut best_tour = population[0].clone();
    let mut best_length = calculate_tour_length(&best_tour, graph);

    for tour in &population {
        let length = calculate_tour_length(tour, graph);
        if length < best_length {
            best_tour = tour.clone();
            best_length = length;
        }
    }

    best_tour
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::{create_distance_matrix, Point};
    use std::f64::consts::PI;

    #[test]
    fn test_genetic_algorithm_visits_all() {
        let points = vec![
            Point { x: 0.0, y: 0.0, angle: PI, id: 0 },
            Point { x: 1.0, y: 0.0, angle: 0.0, id: 1 },
            Point { x: 1.0, y: 1.0, angle: PI / 4.0, id: 2 },
            Point { x: 0.0, y: 1.0, angle: PI / 2.0, id: 3 },
        ];
        let graph = create_distance_matrix(&points);
        let tour = optimize(&graph, 4, 50, 100, 0.1);

        assert_eq!(tour.len(), points.len());

        // Check no duplicates
        let mut seen = std::collections::HashSet::new();
        for &id in &tour {
            assert!(seen.insert(id));
        }
    }
}
